Mathlib/MeasureTheory/Measure/SeparableMeasure.lean

```diff
@@ -0,0 +1,486 @@
+    · exact fun n ↦ ne_top_of_le_ne_top hμs (measure_mono (inter_subset_right ..))
+    · exact ne_top_of_le_ne_top hμs
+        (measure_mono (iUnion_subset (fun i ↦ inter_subset_right ..)))
+
+end MeasureDense
+
+section IsSeparable
+
+/-! ### Definition of a separable measure space, sufficient condition -/
+
+/-- A measure `μ` is separable if there exists a countable and measure-dense family of sets.
+
+The term "separable" is justified by the fact that the definition translates to the usual notion
+of separability in the metric space made by constant indicators equipped with the `L¹` norm. -/
+class IsSeparable (μ : Measure X) : Prop :=
+  exists_countable_measureDense : ∃ 𝒜, 𝒜.Countable ∧ μ.MeasureDense 𝒜
+
+/-- By definition, a separable measure admits a countable and measure-dense family of sets. -/
+theorem exists_countable_measureDense [IsSeparable μ] :
+    ∃ 𝒜, 𝒜.Countable ∧ μ.MeasureDense 𝒜 :=
+  IsSeparable.exists_countable_measureDense
+
+/-- If a measurable space is countably generated and equipped with a `σ`-finite measure, then the
+measure is separable. This is not an instance because it is used below to prove the more
+general case where `μ` is s-finite. -/
+theorem isSeparable_of_sigmaFinite [CountablyGenerated X] [SigmaFinite μ] :
+    IsSeparable μ where
+  exists_countable_measureDense := by
+    have h := countable_countableGeneratingSet (α := X)
+    have hgen := generateFrom_countableGeneratingSet (α := X)
+    let 𝒜 := (countableGeneratingSet X) ∪ {μ.toFiniteSpanningSetsIn.set n | n : ℕ}
+    have count_𝒜 : 𝒜.Countable :=
+      countable_union.2 ⟨h, countable_iff_exists_subset_range.2
+        ⟨μ.toFiniteSpanningSetsIn.set, fun _ hx ↦ hx⟩⟩
+    refine ⟨generateSetAlgebra 𝒜, countable_generateSetAlgebra count_𝒜,
+      measureDense_of_generateFrom_setAglebra_of_sigmaFinite isSetAlgebra_generateSetAlgebra
+      { set := μ.toFiniteSpanningSetsIn.set
+        set_mem := fun n ↦ self_subset_generateSetAlgebra (𝒜 := 𝒜) <| Or.inr ⟨n, rfl⟩
+        finite := μ.toFiniteSpanningSetsIn.finite
+        spanning := μ.toFiniteSpanningSetsIn.spanning }
+      (le_antisymm ?_ (generateFrom_le (fun s hs ↦ ?_)))⟩
+    · rw [← hgen]
+      exact generateFrom_mono <| le_trans self_subset_generateSetAlgebra <|
+        generateSetAlgebra_mono <| subset_union_left ..
+    · induction hs with
+      | base t t_mem =>
+        rcases t_mem with t_mem | ⟨n, rfl⟩
+        · exact hgen ▸ measurableSet_generateFrom t_mem
+        · exact μ.toFiniteSpanningSetsIn.set_mem n
+      | empty => exact MeasurableSet.empty
+      | compl t _ t_mem => exact MeasurableSet.compl t_mem
+      | union t u _ _ t_mem u_mem => exact MeasurableSet.union t_mem u_mem
+
+/-- If a measurable space is countably generated and equipped with an s-finite measure, then the
+measure is separable. -/
+instance [CountablyGenerated X] [SFinite μ] : IsSeparable μ where
+  exists_countable_measureDense := by
```

Comment:
I haven't tried to do this myself yet, but it feels to me like there should be some sort of more general statement that is easier to prove. Maybe something like a (countable) sum of separable measures is separable (is that true?); if so, then that would imply this easily.

Reply:
I tried to prove that a sum of separable measures is separable when @sgouezel suggested to generalize from sigma-finite to s-finite measures, but failed spectacularly, and so went for this less direct proof. I did not even manage to prove that the sum of two separable measures is separable. I do not claim it is false since I wasn't able to find a counter example (this would involve separable measures which are not s-finite and I'm struggling to find one).

Reply:
Here is a sketch of proof that the sum of two separable measures is separable (I haven't checked details, so this may be nonsense).

First, if mu is separable, consider the union `A mu` of the finite-measure sets in a countable measure-dense subset. On the set `A mu`, the measure `mu` is sigma-finite. And in the complement of `A mu`, every set has measure zero or infinity (otherwise, you could approximate such a set by a set in the measure-dense subset, which would then not be contained in `A mu`, contradiction). Moreover, the sigma-algebra on `A mu` is countably generated.

Suppose now mu and nu are separable. Then, on `A mu \cap A nu`, the measure mu + nu is sigma-finite, and the algebra is countably generated, so it is separable there. And on its complement, every set has measure zero or infinite, so won't be relevant for the approximation property.

(Note that this definitely isn't simpler than the current argument, so it doesn't really answer Jireh's question of a simple proof).

Reply:
However, taking inspiration on this, if we assume that the sigma-algebra is compactly generated, I think it shows that separable measures are exactly the ones which can be written as the sum of a sigma-finite measure and a 0-infinity measure.

Reply:
1. Morally, up to measure zero, everything is generated by intersections of sets in the two measure-dense countable sets, so it should be countably generated (well, this depends on the precise definition of countably generated, but I'm not expecting a real difficulty here).
2. You're absolutely right, I had overlooked this one. Here is a possible way out to deal with `(A mu) \ (A nu)`. Restricting to one of the sets in the measure-dense family, I will assume that mu is finite on `A mu`. Then I claim that there is a subset `B mu nu` of `A mu` on which `nu` is zero, and such that on its complement `mu + nu` is a zero-infinity measure. It's an instance of the Hahn decomposition, but simpler here as `nu` is already zero-infinity: just consider a subset of `A mu` of maximal `mu`-measure on which `nu` is zero. Then, on `B mu nu`, the measure `nu` is zero so one can use the measure-dense set of mu (intersected with `B mu nu`). While on `A mu \ B mu nu`, the measure `mu + nu` is zero-infinity, so there is nothing to be approximated.

This is getting more and more complicated, so it's probably not a good idea to try to formalize this since the version you have is already very satisfactory!